### ISSUE TITLE
crash report: storage.cpp line 925 

### ISSUE THREAD

**Issue by WaterSibilantFalling** [arvidn/libtorrent#2313]
**Please provide the following information**

libtorrent version (or branch): 1.1.4
platform/architecture:  Linux 4.13.0-rc5-686-pae \#1 SMP Debian 4.13\~rc5-1\~exp1 (2017-08-14) i686 GNU/Linux
from config.log, compiler and compiler version:
configure:2561: checking for gcc
configure:2577: found /usr/bin/gcc
configure:2588: result: gcc
configure:2817: checking for C compiler version
configure:2826: gcc --version >&5
gcc (Debian 6.3.0-18) 6.3.0 20170516
libtorrent was being used in  qBittorrent. 

I have built libtorrent twice, once a couple of weeks ago, and then again before that another couple of weeks. I git clone https://github.com/arvidn/libtorrent.git and then git checkout "R_1_1". 

For the last month or so, libtorrent, and thus qbittorrent, has been crashing every hour or so. Before that it would never crash, or once every couple of months if I was unlucky.

Here is the crash message:

``` 
Received a tracker reply from http://www.metal-tracker.com:2710/b9f616bc3a78f84da819e34a5c3a98fb/announce (Num_peers = 9)
Received a tracker reply from http://www.metal-tracker.com:2710/b9f616bc3a78f84da819e34a5c3a98fb/announce (Num_peers = 1)
Received a tracker reply from http://www.metal-tracker.com:2710/b9f616bc3a78f84da819e34a5c3a98fb/announce (Num_peers = 0)
Saving resume data in "/home/me/.local/share/data/qBittorrent/BT_backup/c938f452bb17edc06bd21e6101abd4cf8b25523d.fastresume"
Saving resume data in "/home/me/.local/share/data/qBittorrent/BT_backup/66d2206956ef9914dc9f0c499437b8ced9cedc5e.fastresume"
Saving resume data in "/home/me/.local/share/data/qBittorrent/BT_backup/dd8a710e42587e8af30d58d9ec153dbcc10d4100.fastresume"
Saving resume data in "/home/me/.local/share/data/qBittorrent/BT_backup/7cf1934c37f75f5ac24301a571ea7d93905f5e66.fastresume"
assertion failed. Please file a bugreport at https://github.com/arvidn/libtorrent/issues
Please include the following information:

version: 1.1.4.0-20c1407

file: 'storage.cpp'
line: 925
function: virtual void libtorrent::default_storage::write_resume_data(libtorrent::entry&, libtorrent::storage_error&) const
expression: s.file_size == file_size

stack:
1: assert_fail(char const*, int, char const*, char const*, char const*, int)
2: libtorrent::default_storage::write_resume_data(libtorrent::entry&, libtorrent::storage_error&) const
3: 
4: 
5: 
6: 
7: 
8: 
9: 
10: clone

Catching signal: SIGINT
Exiting cleanly


*************************************************************
Catching signal: SIGABRT
Please file a bug report at http://bug.qbittorrent.org and provide the following information:

qBittorrent version: v3.4.0beta
stack trace:
Saving resume data in "/home/me/.local/share/data/qBittorrent/BT_backup/ce281496165bde182de08c0b9e34caa48d95c5fe.fastresume"
  linux-gate.so.1 : __kernel_vsyscall()+0x9  [0xb7f17ce9]
  /lib/i386-linux-gnu/libc.so.6 : gsignal()+0xb0  [0xb6f95dc0]
  /lib/i386-linux-gnu/libc.so.6 : abort()+0x157  [0xb6f97287]
  /usr/lib/i386-linux-gnu/libtorrent-rasterbar.so.9 : assert_fail(char const*, int, char const*, char const*, char const*, int)+0x94  [0xb7a7a1a4]
  /usr/lib/i386-linux-gnu/libtorrent-rasterbar.so.9 : libtorrent::default_storage::write_resume_data(libtorrent::entry&, libtorrent::storage_error&) const+0x160  [0xb7c74280]
  /usr/lib/i386-linux-gnu/libtorrent-rasterbar.so.9 : ()+0xc015c  [0xb7ac515c]
  /usr/lib/i386-linux-gnu/libtorrent-rasterbar.so.9 : ()+0xc0ae9  [0xb7ac5ae9]
  /usr/lib/i386-linux-gnu/libtorrent-rasterbar.so.9 : ()+0xc0f00  [0xb7ac5f00]
  /usr/lib/i386-linux-gnu/libtorrent-rasterbar.so.9 : ()+0xc724d  [0xb7acc24d]
  /usr/lib/i386-linux-gnu/libtorrent-rasterbar.so.9 : ()+0xc8996  [0xb7acd996]
  /usr/lib/i386-linux-gnu/libtorrent-rasterbar.so.9 : ()+0xc84fd  [0xb7acd4fd]
  /lib/i386-linux-gnu/libpthread.so.0 : ()+0x627a  [0xb731727a]
  /lib/i386-linux-gnu/libc.so.6 : clone()+0x66  [0xb7051b56]
Aborted
````

when I then cat the file that libtorrent was writing, it looks like there is some weirness in it: 

```
 cat /home/me/.local/share/data/qBittorrent/BT_backup/ce281496165bde182de08c0b9e34caa48d95c5fe.fastresume
d11:active_timei1385387e10:added_timei1496498219e15:announce_to_dhti1e15:announce_to_lsdi1e20:announce_to_trackersi1e12:auto_managedi1e12:banned_peers0:13:banned_peers60:16:blocks per piecei4e14:completed_timei1496500140e19:download_rate_limiti-1e10:file sizeslli7160681ei1496500139eeli6625688ei1496500139eeli5545271ei1496500139eeli5300757ei1496500135eeli5191040ei1496500139eeli4870257ei1496500139eeli4845178ei1496500140eeli4815924ei1496500139eeli4155545ei1496500139eeli4085543ei1496500139eeli2675971ei1496500139eeli211708ei1496500134eee11:file-format22:libtorrent resume file12:file-versioni1e13:finished_timei1378832e9:info-hash20:Î([Þ-à
                                                                                                                                                                                         4ʤÅþ18:last_seen_completei1496500231e18:libtorrent-version7:1.1.4.012:mapped_filesl46:Too Lat..... 

```
The  "gap"  in that fast-resume file is kinda missing due to the github formatting, but there is some weird, non-ascii characters in that fastresume file, like this: 

```
:info-hash20:Î([Þ-à
                                                                                                                                                                                         4ʤÅþ18:last_see

```

Anyway, qbittorrent and libtorrent are crashing. 

Thanks for any help with this.

**Comment by arvidn** [comment]
the fastresume file is bencoded, which includes binary parts (i.e. non-printable characters), so that's not too surprising.

However, this assert was put in a while ago to try to tease out a bug where the resume data wasn't saved correctly. libtorrent cache the sizes of files on disk and clears the cache when writing to them or force-rechecking. This assertion failure suggests that a file on disk mismatches the cached size for that file, which is unexpected. It's not impossible though, if 3rd party applications modify files "under the feet" of libtorrent. Do you know if you have any 3rd party applications modifying files? I believe windows media player normally does that, while adding id3 tags.

A related issue involving force-rechecking of torrents was fixed here, do you know if you have this change in your build?

https://github.com/arvidn/libtorrent/commit/1f3730fa5ad1658d8ca5f100cbc87b80938d950e#diff-a85f9d12910f601cfa0d67cbe645f968

**Comment by AbsurdMagpieScrutinies** [comment]
> the fastresume file is bencoded, which includes binary parts (i.e. non-printable characters), so that's not too surprising.

OK, so whether or not that fastresume file was corrupt, the handling of fastresume files can handle arbitrary non-ascii. control characters. So it wasn't the fastresume file. 

> Do you know if you have any 3rd party applications modifying files? 

No, there isn't, and that content hadn't been touched since it was downloaded (like many people I have GBs of unprocessed/unconsumed content that i must get around to). I would have thought that if any change or corruption was made to the content represented by a *.torrent file then, upon recheck, the torrenting library or app would see the torrent's content as uncomplete - due to hash mis-matches - and then try to download those changed parts of the torrent's content again.  

> A related issue involving force-rechecking of torrents was fixed here, do you know if you have this change in your build?
>
> https://github.com/arvidn/libtorrent/commit/1f3730fa5ad1658d8ca5f100cbc87b80938d950e#diff-a85f9d12910f601cfa0d67cbe645f968

That change was committed on Aug 6th and I built libtorrent on Sept 1 so that change would have been incorporated, I guess.

**Comment by AbsurdMagpieScrutinies** [comment]
Would you like the *.fastresume files upon which libtorrent crashes?
